plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c

```diff
@@ -0,0 +1,510 @@
+
+	ret = avcodec_encode_video2(enc->context, &av_pkt, hwframe,
+			&got_packet);
+	if (ret < 0) {
+		warn("vaapi_encode: Error encoding: %s", av_err2str(ret));
+		return false;
+	}
+
+	if (got_packet && av_pkt.size) {
+		if (enc->first_packet) {
+			uint8_t *new_packet;
+			size_t size;
+
+			enc->first_packet = false;
+			obs_extract_avc_headers(av_pkt.data, av_pkt.size,
+					&new_packet, &size,
+					&enc->header, &enc->header_size,
+					&enc->sei, &enc->sei_size);
+
+			da_copy_array(enc->buffer, new_packet, size);
+			bfree(new_packet);
+		} else {
+			da_copy_array(enc->buffer, av_pkt.data, av_pkt.size);
+		}
+
+		packet->pts = av_pkt.pts;
+		packet->dts = av_pkt.dts;
+		packet->data = enc->buffer.array;
+		packet->size = enc->buffer.num;
+		packet->type = OBS_ENCODER_VIDEO;
+		packet->keyframe = obs_avc_keyframe(packet->data, packet->size);
+		*received_packet = true;
+	} else {
+		*received_packet = false;
+	}
+
+	av_free_packet(&av_pkt);
+	av_frame_free(&hwframe);
+	return true;
+}
+
+static void set_visible(obs_properties_t *ppts, const char *name, bool visible)
+{
+	obs_property_t *p = obs_properties_get(ppts, name);
+	obs_property_set_visible(p, visible);
+}
+
+
+static void vaapi_defaults(obs_data_t *settings)
+{
+	obs_data_set_default_string(settings, "vaapi_device", "/dev/dri/renderD128");
+	obs_data_set_default_int(settings, "vaapi_codec", AV_CODEC_ID_H264);
+	obs_data_set_default_int(settings, "profile", 578);
+	obs_data_set_default_int(settings, "level", 40);
+	obs_data_set_default_int(settings, "bitrate", 2500);
+	obs_data_set_default_int(settings, "keyint_sec", 0);
+	obs_data_set_default_int(settings, "bf", 0);
+	obs_data_set_default_int(settings, "qp", 20);
+	obs_data_set_default_int(settings, "quality", 0);
+	obs_data_set_default_int(settings, "rendermode", 0);
+}
+
+static obs_properties_t *vaapi_properties(void *unused)
+{
+	UNUSED_PARAMETER(unused);
+
+	obs_properties_t *props = obs_properties_create();
+	obs_property_t *list;
+
+	list = obs_properties_add_list(props, "vaapi_device", "VAAPI Device",
+			OBS_COMBO_TYPE_LIST, OBS_COMBO_FORMAT_STRING);
+	char path[128] = "/dev/dri/renderD1";
+	for (int i = 28;; i++) {
+		sprintf(path, "/dev/dri/renderD1%d", i);
+		if (access(path, F_OK) == 0) {
+			char card[128] = "Card: ";
+			sprintf(card, "Card%d: %s", i-28, path);
+			obs_property_list_add_string(list, card, path);
+		} else {
+			break;
+		}
+	}
+
+	list = obs_properties_add_list(props, "vaapi_codec", "VAAPI Codec",
+			OBS_COMBO_TYPE_LIST, OBS_COMBO_FORMAT_INT);
+
+	obs_property_list_add_int(list, "H.264 (default)", AV_CODEC_ID_H264);
+
+	list = obs_properties_add_list(props, "level", "Level",
+			OBS_COMBO_TYPE_LIST, OBS_COMBO_FORMAT_INT);
+	obs_property_list_add_int(list, "480p30 (3.0)", 30);
+	obs_property_list_add_int(list, "720p30/480p60  (3.1)", 31);
+	obs_property_list_add_int(list, "Compatibility mode  (4.0 default)", 40);
+	obs_property_list_add_int(list, "720p60/1080p30 (4.1)", 41);
```

Comment:
I'm finding mixed information on the maximum res/FPS for Level 4.0 and 4.1. I've seen 2,048×1,024@30 and 1920×1088@30.1 for both 4.0 and 4.1.

Reply:
Levels are mostly used for documenting hardware decoder capabilities and are not defined in terms of resolutions or framerates. Devices which support certain levels may choose not to support resolution and frame rates that theoretically would fit in that level. The wiki page is a handy reference https://en.wikipedia.org/wiki/H.264/MPEG-4_AVC#Levels

Reviewing the code this sets a level parameter on the ffmpeg encoder which probably forces the stream to be fall within the level constraints. I would need to review the previous thread to pin down why level constraints were used if there is a reason at all.

Reply:
Yes, I've read that Wikipedia page quite often.  I'm aware that the Levels actually specify something different than resolution/FPS (macroblocks/second, macroblocks/frame, bitrate, buffer sizes, etc.), though those specifications can be converted to expected allowable resolution/FPS.  I've also read other resources on the supposed resolution/FPS limits.

- https://www.vocal.com/video/profiles-and-levels-in-h-264-avc/
- http://blog.mediacoderhq.com/h264-profiles-and-levels/
- https://www.itu.int/ITU-T/worksem/h325/200605/presentations/s3p1-luthra.pdf

My point here was that I'd found conflicting information on expected maximums for resolution/FPS, so I wasn't sure what was appropriate to cite in this list of options.  Level 4.0 and Level 4.1 differ mostly in maximum bitrate, which allows for more reference frames, I believe.  Admittedly, my experience and knowledge with this area is limited.

Reply:
Most modern hardware should handle 5.1, yes.  The presentation I linked earlier was from 2006 and mentioned Level 5.1.  There are even levels above that, which modern hardware supports.  For other encoders, we leave it at auto/0, and let the encoder/FFmpeg decide, which we usually recommend users to leave it at because selecting a level too low for their desired resolution/FPS/bitrate can cause the encoder to fail to start, which is a commonly seen support case.

I'm not sure _why_ we allow users to select Level either, since it mostly leads to misconfigurations.  I just know that we do allow it in Advanced Output Mode for other encoders.

Reply:
It may be related to the constrained baseline profile choice which was presumably because twitch ingest fails on other profiles. But twitch's own recommendations are a range of levels up to 4.2 and profiles of main/high so more information on why AMD devices at main/high were not working appropriately and if it still affects them or any other devices.

Reply:
Yeah, Twitch's current recommendations are main/high, as seen on [stream.twitch.tv](https://stream.twitch.tv/).  Their [older recommendations](https://help.twitch.tv/customer/en/portal/articles/1253460-broadcast-requirements), which are still publicly available, _do_ say "Main (preferred) or Baseline".  Not sure why "main" would fail these days, but more testing may help.